### PR TITLE
BigQuery: Use DatasetListItem for client.list_datasets

### LINKED ISSUE #4373: BigQuery Python Client v0.28.0: num_rows, schema property of a Table returning None and empty list
Hi,

I am using the BigQuery Python client v0.28.0

Using the bigquery client to list all tables in a dataset using method `list_dataset_tables`, I get an iterator of table references in the dataset. When I try to access the `num_rows` and `schema` properties of the Table object it returns `None` and `[]`.

But when I access the table using `client.get_table()` I am able to access both the num_rows and schema properties.

I am confused since both the methods return a `google.cloud.bigquery.table.Table` object but are inconsistent.

```
for table in bqclient.list_dataset_tables(dataset_ref):
	print(table.num_rows)
	print(table.schema)
```
Returns None and []
```
table = bqclient.get_table(table)
	print(table.num_rows)
	print(table.schema)
```
Returns the number of rows and list of schemaFields

Thanks

### PR DESCRIPTION
**Pull request by tswast**
Listing datasets only includes a subset of the properties available on a
dataset. The DatasetListItem class is used to explicitly document which
features are available and to prevent confusion from trying to use the
resulting object in other contexts, like updating.

Closes #4373.

### CODE REVIEW

**Comment by tswast** [comment]
Ping. Please review.

File: bigquery/google/cloud/bigquery/dataset.py [hunk 1]
```diff
@@ -546,4 +545,91 @@ def table(self, table_id):
...
+        self._properties = resource
+
+    @property
+    def project(self):
+        """Project bound to the dataset.
+
+        :rtype: str
+        :returns: the project.
+        """
+        return self._properties.get('datasetReference').get('projectId')
```
Comment (tseaver) [inline bigquery/google/cloud/bigquery/dataset.py]: Presumes that the `datasetReference` key is always populated appropriately (with at least an empty dict).

We've not been perfectly consistent about defending against "perverse" back-end responses, so maybe this one can go by.  In that case, is `self._properties['datasetReference']['projectId']` clearer?

If we *do* leave the `.get()` bits in place, then we need to exercise the implicit branches for missing `datasetReference` in the unit tests.

File: bigquery/google/cloud/bigquery/dataset.py [hunk 2]
```diff
@@ -546,4 +545,91 @@ def table(self, table_id):
...
+
+    @property
+    def dataset_id(self):
+        """Dataset ID.
+
+        :rtype: str
+        :returns: the dataset ID.
+        """
+        return self._properties.get('datasetReference').get('datasetId')
```
Comment (tseaver) [inline bigquery/google/cloud/bigquery/dataset.py]: Ditto here.

File: bigquery/google/cloud/bigquery/dataset.py [hunk 3]
```diff
@@ -546,4 +545,91 @@ def table(self, table_id):
...
+        return self._properties.get('friendlyName')
+
+    @property
+    def labels(self):
+        """Labels for the dataset.
+
+        :rtype: dict, {str -> str}
+        :returns: A dict of the the dataset's labels.
+        """
+        return self._properties['labels']
```
Comment (tseaver) [inline bigquery/google/cloud/bigquery/dataset.py]: This one is *not* defensive against missing labels:  I would say that one is *more* likely to be missing than `datasetReference`, based on past experience.

If we mean to raise `KeyError` from the property, then we should test that as well.

Comment (tswast) [reply bigquery/google/cloud/bigquery/dataset.py]: Good point. Agreed that it's more likely to be missing.

File: bigquery/google/cloud/bigquery/dataset.py (hunk 1 above)
Comment (tswast) [reply bigquery/google/cloud/bigquery/dataset.py]: I agree that `self._properties['datasetReference']['projectId']` is clearer in this case. I've added an explicit check for these properties in the constructor.

**Pull request merged**